src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -11251,14 +11251,24 @@ private bool ScanCast(bool forPattern = false)
                     // following a cast.
                     return !forPattern || this.CurrentToken.Kind switch
                     {
-                        SyntaxKind.PlusToken => true,
-                        SyntaxKind.MinusToken => true,
-                        SyntaxKind.AmpersandToken => true,
-                        SyntaxKind.AsteriskToken => true,
+                        SyntaxKind.PlusToken or
+                        SyntaxKind.MinusToken or
+                        SyntaxKind.AmpersandToken or
+                        SyntaxKind.AsteriskToken or
                         SyntaxKind.DotDotToken => true,
-                        _ => CanFollowCast(this.CurrentToken.Kind)
+                        // If the parenthesized type is followed by a binary pattern, it is not a cast.
+                        SyntaxKind.IdentifierToken when isBinaryPattern() => false,
+                        var tk => CanFollowCast(tk)
                     };
 
+                    bool isBinaryPattern()
+                    {
+                        if (this.CurrentToken.ContextualKind is not (SyntaxKind.OrKeyword or SyntaxKind.AndKeyword))
+                            return false;
+                        EatToken();
+                        return IsPossibleSubpatternElement();
```

Comment:
minor point: if we know we're parsing a pattern, wouldn't a more reasonable error recovery for a scenario like `(int) and` be to say: "the right-hand subpattern is missing", versus saying, "this is a cast and a pattern can't be a cast"?

Reply:
That's already valid code and we should keep accepting that. See CastExpressionInPattern_02

Looks like `x is (int)or or y` works too. Will add another test for that. Any other scenarios you think we should test for?

Reply:
Thanks, it didn't occur to me that a cast expression can still be constant-valued and therefore is a valid constant pattern.

I'll try and make a pass over the tests and let you know if I come up with additional scenarios.